lib/semantics/check-declarations.cc

```diff
@@ -889,12 +905,39 @@ void CheckHelper::Check(const Scope &scope) {
   } else if (scope.IsDerivedType()) {
     return;  // PDT instantiations have null symbol()
   }
+  for (const auto &set : scope.equivalenceSets()) {
+    CheckEquivalenceSet(set);
+  }
   for (const auto &pair : scope) {
     Check(*pair.second);
   }
   for (const Scope &child : scope.children()) {
     Check(child);
   }
+  if (scope.kind() == Scope::Kind::BlockData) {
+    CheckBlockData(scope);
+  }
+}
+
+void CheckHelper::CheckEquivalenceSet(const EquivalenceSet &) {
+  // TODO: Move C8106 (&al.) checks here from resolve-names-utils.cc
+}
+
+void CheckHelper::CheckBlockData(const Scope &scope) {
+  // BLOCK DATA subprograms should contain only named common blocks.
+  for (const auto &pair : scope) {
+    const Symbol &symbol{*pair.second};
+    if (!(symbol.has<CommonBlockDetails>() || symbol.has<UseDetails>() ||
+            symbol.has<UseErrorDetails>() || symbol.has<DerivedTypeDetails>() ||
+            symbol.has<SubprogramDetails>() ||
+            symbol.has<ObjectEntityDetails>() ||
+            (symbol.has<ProcEntityDetails>() &&
+                !symbol.attrs().test(Attr::POINTER)))) {
+      messages_.Say(symbol.name(),
+          "'%s' may not appear in a BLOCK DATA subprogram"_err_en_US,
+          symbol.name());
```

Comment:
There are a bunch of declaration-constructs that are prohibited by C1415 and I don't think they can all be detected by looking at the symbols. import-stmt and access-stmt are caught in `resolve-names.cc` but there are more like that.

Reply:
So long as nothing in the `BLOCK DATA` generates any code or any allocates any storage not in `COMMON`, I don't think that we should care.

Reply:
I think we want, at least optionally, a strict standard-checking mode. But maybe it's not very high priority.

Reply:
If there's a todo here, please add a TODO.

Reply:
Yes, I believe it would be nice to have a TODO of what is not yet implemented

I have tried this small code with block data and it is passing. Is that correct? Shouldn't the compiler print an error when using External and BIND?

block data  bd1_c1416
    integer, external :: beta
end block data bd1_c1416

block data bd2_c1416
  integer, bind(C) :: x
end block data bd2_c1416

In my mind this could be implemented inside resolve-names.cc in by checking for 
(Attr::EXTERNAL) and (Attr::BIND_C) when inside BlockData no?